util/patch/patch_test.go

```diff
@@ -17,232 +17,438 @@ limitations under the License.
+			}, timeout).Should(BeTrue())
+		})
+	})
+
+	It("Should patch a corev1.Node object conditions", func() {
+		now := metav1.Now().Rfc3339Copy()
+
+		obj := &corev1.Node{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node1",
 			},
-			after: &clusterv1.Cluster{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      "test-cluster",
-					Namespace: "test-namespace",
-					Finalizers: []string{
-						clusterv1.ClusterFinalizer,
-					},
-				},
+		}
+
+		By("Creating a Node object")
+		Expect(testEnv.Create(ctx, obj)).ToNot(HaveOccurred())
+		key := client.ObjectKey{Name: obj.Name}
+		defer func() {
+			Expect(testEnv.Delete(ctx, obj)).To(Succeed())
+		}()
+
+		By("Creating a new patch helper")
+		patcher, err := NewHelper(obj, testEnv)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Appending a new Ready condition")
+		obj.Status.Conditions = append(obj.Status.Conditions, corev1.NodeCondition{
+			Type:               "CustomCondition",
+			Status:             corev1.ConditionTrue,
+			LastHeartbeatTime:  now,
+			LastTransitionTime: now,
+			Reason:             "reason",
+			Message:            "message",
+		})
+
+		By("Patching the Node")
+		Expect(patcher.Patch(ctx, obj)).To(Succeed())
+
+		By("Validating the object has been updated")
+		Eventually(func() bool {
+			objAfter := obj.DeepCopy()
+			if err := testEnv.Get(ctx, key, objAfter); err != nil {
+				return false
+			}
+
+			return reflect.DeepEqual(objAfter.Status.Conditions, obj.Status.Conditions)
+		}, timeout).Should(BeTrue())
+	})
+
+	Describe("Should patch a clusterv1.Cluster", func() {
+		obj := &clusterv1.Cluster{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "test-",
+				Namespace:    "default",
 			},
-		},
-		{
-			name: "Only add ownerref update to unstructured object",
-			before: &unstructured.Unstructured{
-				Object: map[string]interface{}{
-					"kind":       "BootstrapConfig",
-					"apiVersion": "bootstrap.cluster.x-k8s.io/v1alpha3",
-					"metadata": map[string]interface{}{
-						"name":      "test-bootstrap",
-						"namespace": "default",
-					},
-				},
+		}
+
+		Specify("add a finalizers", func() {
+			obj := obj.DeepCopy()
+
+			By("Creating the object")
+			Expect(testEnv.Create(ctx, obj)).ToNot(HaveOccurred())
+			key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+			defer func() {
+				Expect(testEnv.Delete(ctx, obj)).To(Succeed())
+			}()
+
+			By("Creating a new patch helper")
+			patcher, err := NewHelper(obj, testEnv)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Adding a finalizer")
+			obj.Finalizers = append(obj.Finalizers, clusterv1.ClusterFinalizer)
+
+			By("Patching the object")
+			Expect(patcher.Patch(ctx, obj)).To(Succeed())
+
+			By("Validating the object has been updated")
+			Eventually(func() bool {
+				objAfter := obj.DeepCopy()
+				if err := testEnv.Get(ctx, key, objAfter); err != nil {
+					return false
+				}
+
+				return reflect.DeepEqual(obj.Finalizers, objAfter.Finalizers)
```

Comment:
Would it be better to compare the finalizers from the state of the finalizers before `Patch` is called on the object? As in copying them out to a var. I'm wondering if there is a scenario where `Patch` might not actually update the API with the new finalizer and overwrite the finalizers on obj

```
finalizers := append(obj.Finalizers, clusterv1.ClusterFinalizer)
obj.Finalizers = finalizers
// Compare against `finalizers` instead of `obj.Finalizers`
```

Reply:
Not sure if I'm following, the `obj` value is always set to what's before the patch is sent, the object should never be updated, we have internal copies for that

Reply:
Never mind, in that case this is fine 👍